osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs

```diff
@@ -57,6 +57,12 @@ protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject,
                     {
                         hitObject.FadeOut(duration);
 
+                        // Keep updating the object even after it stopped being visible.
+                        // This is required because of the custom logic in DrawableHit's Update method.
+                        // Specifically, pressHandledThisFrame wasn't reset, which caused further hits
+                        // within the object's lifetime to be rejected.
+                        hitObject.AlwaysPresent = true;
```

Comment:
Not sure how to feel about this. We generally consider `AlwaysPresent` to be the tool of last resort because using it can have silent adverse performance effects. And while yes this is constrained to the hidden mod, I'd rather explore alternative solutions.

One potential solution that I can think of is to remove the reliance of the hit detection logic on the update loop by storing the exact time of the last press instead:

```diff
diff --git a/osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs b/osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
index ad17955f75..2c3b4a8d18 100644
--- a/osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
+++ b/osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
@@ -57,12 +57,6 @@ protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject,
                     {
                         hitObject.FadeOut(duration);
 
-                        // Keep updating the object even after it stopped being visible.
-                        // This is required because of the custom logic in DrawableHit's Update method.
-                        // Specifically, pressHandledThisFrame wasn't reset, which caused further hits
-                        // within the object's lifetime to be rejected.
-                        hitObject.AlwaysPresent = true;
-
                         // DrawableHitObject sets LifetimeEnd to LatestTransformEndTime if it isn't manually changed.
                         // in order for the object to not be killed before its actual end time (as the latest transform ends earlier), set lifetime end explicitly.
                         hitObject.LifetimeEnd = state == ArmedState.Idle || !hitObject.AllJudged
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs
index 44225ab289..1ef426854e 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs
@@ -35,7 +35,7 @@ public partial class DrawableHit : DrawableTaikoStrongableHitObject<Hit, Hit.Str
 
         private bool validActionPressed;
 
-        private bool pressHandledThisFrame;
+        private double? lastPressHandleTime;
 
         private readonly Bindable<HitType> type = new Bindable<HitType>();
 
@@ -76,7 +76,8 @@ protected override void OnFree()
 
             HitActions = null;
             HitAction = null;
-            validActionPressed = pressHandledThisFrame = false;
+            validActionPressed = false;
+            lastPressHandleTime = null;
         }
 
         private void updateActionsFromType()
@@ -114,7 +115,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
 
         public override bool OnPressed(KeyBindingPressEvent<TaikoAction> e)
         {
-            if (pressHandledThisFrame)
+            if (lastPressHandleTime == Time.Current)
                 return true;
             if (Judged)
                 return false;
@@ -128,7 +129,7 @@ public override bool OnPressed(KeyBindingPressEvent<TaikoAction> e)
 
             // Regardless of whether we've hit or not, any secondary key presses in the same frame should be discarded
             // E.g. hitting a non-strong centre as a strong should not fall through and perform a hit on the next note
-            pressHandledThisFrame = true;
+            lastPressHandleTime = Time.Current;
             return result;
         }
 
@@ -139,15 +140,6 @@ public override void OnReleased(KeyBindingReleaseEvent<TaikoAction> e)
             base.OnReleased(e);
         }
 
-        protected override void Update()
-        {
-            base.Update();
-
-            // The input manager processes all input prior to us updating, so this is the perfect time
-            // for us to remove the extra press blocking, before input is handled in the next frame
-            pressHandledThisFrame = false;
-        }
-
         protected override void UpdateHitStateTransforms(ArmedState state)
         {
             Debug.Assert(HitObject.HitWindows != null);

```

@ppy/team-client thoughts?

Reply:
just to advance this PR quicker since it's pretty important - is it fine if I just apply the patch you sent?

Reply:
If I wanted to, I could have applied it myself. The only reason I haven't done it is that I collect more opinions on it. This change is relatively important, but is not hotfix tier either. It can wait for someone to take a look at it.

Reply:
I prefer the proposed diff. The only worry I have is that there might be a weird case where `Time` does not change between two frames. I don't know if this can theoretically happen (check whether it can is probably far too hard, considering it's probably an edge case involving several levels of clocks and platform specific timing quirks)...

*that said*, in such a case it's probably fine to handle as-is. so i'd be okay with this.

Reply:
Applied in https://github.com/ppy/osu/pull/24305/commits/7e98d319d85daeab6a5871e397ae102efa8bd2b9, with a test case covering the "subsequent hits on same frame are consumed by the already-hit object" in https://github.com/ppy/osu/pull/24305/commits/5cedd428bb06cbd2e256ef45fe6b5f77e110c43f, just to make sure I haven't broken anything.